src/Data/Validation.hs

```diff
@@ -47,184 +23,274 @@ import Data.Map hiding (null)
-testMatch e a1 a2 = case a1 == a2 of
-  True  -> Nothing
-  False -> Just e
-
-validateField :: Validatable e a b => Name -> a -> VCtx e b
-validateField n a = case validation a of
-  ValidCtx b            -> ValidCtx b
-  DisputedCtx ges les b -> DisputedCtx [] (insert [n] ges $ mapKeys (\k -> [n] ++ k) les) b
-  RefutedCtx ges les    -> RefutedCtx [] (insert [n] ges $ mapKeys (\k -> [n] ++ k) les)
-
-optional :: Maybe a -> (a -> VCtx e b) -> VCtx e (Maybe b)
+-- | Checks that a 'Data.Maybe.Maybe' value is a 'Data.Maybe.Just'.
+-- If not, it adds the given failure to the result and validation end.
+isRequired :: f -> ValueCtx (Maybe a) -> VCtx f (ValueCtx a)
+isRequired f = refuteWith $ \ma -> case ma of
+  Nothing -> Left f
+  Just a  -> Right a
+  
+-- | Checks that the 'Foldable' is empty.
+-- If so, it adds the given failure to the result and validation continues.
+isNull :: (Foldable t) => f -> ValueCtx (t a) -> VCtx f (ValueCtx (t a))
+isNull f = disputeWith $ bool (Just f) Nothing . null
+  
+-- | Checks that the 'Foldable' is not empty.
+-- If not, it adds the given failure to the result and validation continues.
+isNotNull :: (Foldable t) => f -> ValueCtx (t a) -> VCtx f (ValueCtx (t a))
+isNotNull f = disputeWith $ bool (Just f) Nothing . not . null
+
+-- | Checks that a 'IsString' has a length equal to or grater than the given value.
+-- If not, it adds the given failure to the result and validation continues.
+minLength :: (IsString a, Show a) => Int -> f -> ValueCtx a -> VCtx f (ValueCtx a)
+minLength l f = disputeWith $ bool (Just f) Nothing . (<=) l . length . show
+
+-- | Checks that a 'IsString' has a length equal to or less than the given value.
+-- If not, it adds the given failure to the result and validation continues.
+maxLength :: (IsString a, Show a) => Int -> f -> ValueCtx a -> VCtx f (ValueCtx a)
+maxLength l f = disputeWith $ bool (Just f) Nothing . (>=) l . length . show
+
+-- | Checks that a 'Num' is equal to or greater than the given value.
+-- If not, it adds the given failure to the result and validation continues.
+minValue :: (Num a, Ord a) => a -> f -> ValueCtx a -> VCtx f (ValueCtx a)
+minValue l f = disputeWith $ bool (Just f) Nothing . (<=) l
+
+-- | Checks that a 'Num' is equal to or less than the given value.
+-- If not, it adds the given failure to the result and validation continues.
+maxValue :: (Num a, Ord a) => a -> f -> ValueCtx a -> VCtx f (ValueCtx a)
+maxValue l f = disputeWith $ bool (Just f) Nothing . (>=) l
+
+-- | Checks that two fields are equal.
+-- If not, it adds the given failure to the result and validation continues.
+isMatch :: Eq a => f -> VCtx f a -> ValueCtx a -> VCtx f (ValueCtx a)
+isMatch f (ValidCtx a)        = disputeWith (testMatch f a)
+isMatch f (DisputedCtx _ _ a) = disputeWith (testMatch f a)
+isMatch f (RefutedCtx _ _)    = pure
+
+-- | Vaidates a value that implements 'Validatable' and includes any failures under the parent field.

```

Comment:
*Validates
and should this mention something about how it determines the field?

Reply:
Fixed.